### ISSUE TITLE
How are the parameters STFT_WINDOW_LENGTH_SECONDS, STFT_HOP_LENGTH_SECONDS and others parakeets related to each other? 

### ISSUE THREAD

**Issue by tusharpoddar** [tensorflow/models#8567]
I have been working with the vggish_input file to generate the NumPy array for the spectrogram of the input wav file that I select. Instead of getting each window length to be 0.96 seconds, I want each window to be 0.025 seconds long. For this, I change the vggish_params to the following - 

```
STFT_WINDOW_LENGTH_SECONDS = 0.000512

STFT_HOP_LENGTH_SECONDS = 0.010/38.4
NUM_MEL_BINS = NUM_BANDS
MEL_MIN_HZ = 125
MEL_MAX_HZ = 30000
LOG_OFFSET = 0.01  # Offset used for stabilized log of input mel-spectrogram.
EXAMPLE_WINDOW_SECONDS = 0.96/38.4  # Each example contains 96 10ms frames
EXAMPLE_HOP_SECONDS = 0.96/38.4
```

in this, I change STFT_HOP_LENGTH_SECONDS, EXAMPLE_WINDOW_SECONDS and EXAMPLE_HOP_SECONDS in the same proportions. Ie since I wanted to get the hop length to be 0.025 I divided each of them by 38.4. And I changed STFT_WINDOW_LENGTH_SECONDS to 0.000512 in order to get the nperseg(used in the scipy spectrogram function) to be equal to 128 (0.000512 * 2,50,000). Here my sampling frequency is 250,000.  If I dont worry about training the data as yet, will these changed paraments give me the required spectrogram array? 

If not, what am i understanding wrong? 
Thanks in advance.

**Comment by dpwe** [comment]
VGGish takes a spectrogram "patch" of 96 frames x 64 mel channels as input.  This corresponds to 96 successive columns from a spectrogram calculated with 10ms frames (based on 25ms windows), so it covers ~0.96 seconds of input per patch (actually 0.975s if you include the whole of the 25 ms window "at the end").  Your wish for 25ms input window sounds like the 25ms window we already use on the underlying spectrogram, so maybe it's already what you want.

VGGish is a pretrained network.  It was trained on these 96x64 input spectrograms, so if you alter that part of the processing, you'll be running it on inputs that don't resemble the training, so the results will be unpredictable and likely not very useful (it's an embedding, rather than a classifier, so it will still provide a consistent mapping between input and output, but the mapping will not, for instance, help normalize away non-semantic acoustic variations, which is the basic goal of the embedding).

Moreover, if you try to change the input to be a shape other than 96x64, the network can't run at all.

It's possible to get 96 time frames out of 25ms of audio, but it's almost surely not what you want.  Each frame would correspond to ~2.604ms of audio.  At 16 kHz sampling rate, 25ms is 400 samples; you can't divide that integrally by 96, but if you used a hop of 4 samples per frame, you'd get 96 frames in 24ms, which is close.  

If you are happy to vary the sampling rate, then simply increasing the sampling rate to (960/25 * 16000) = 608,000 Hz, and reducing the STFT_*_LENGTH_SECONDS by 25/960 i.e. WINDOW=0.00065104167s and HOP=0.0002604167s should maintain the 400 sample window, 160 sample hop, and 15,600 total sample span per input classification.  But I stress this yields essentially meaningless results and is, as far as I can see, very unlikely to be what you want.

**Comment by tusharpoddar** [comment]
Hi, I completely understand what you are trying to say. I am actually using a sampling frequency of 250000. Actually I am working with the sounds generated in the ocean and your categories do not go with the ones that we want to predict. So right now, we are trying to get the model to produce a spectrogram that gives out the best features of the sounds generated, and then we will try to interpolate that shape to 96 * 64 so that it works with the vggish model. If nothing works, we will then try to train the vggish with our own data. For this, I had a few questions. What all layers is the vggish model currently using? Also, how do I train the model with my own data?

**Comment by dpwe** [comment]
The VGGish architecture is described in [our paper](https://arxiv.org/abs/1609.09430) as simply the [VGG configuration E](https://arxiv.org/abs/1409.1556) classifier with an output layer modified to match the 3087 classes we used to train.

I don't really understand what you're hoping to gain by feeding in sounds from a different domain and with a different sampling rate into the VGGish embedding.  In any case, VGGish doesn't predict any specific categories. I wonder how you will decide if your experiments have "worked" or not.

I'm unable to give you direction on how to train a deep net (I only know how to use our in-house tools) but I'm sure there are lots of tutorials available online. The hardest part will probably be organizing your data and getting enough labels of sufficient quality.

There's nothing particularly special about VGG/VGGish.  We found it worked well for our problem (where we had millions of examples and thousands of distinct labels), but the basic message is that image-classifier-style CNNs can work very well when applied to small patches drawn from auditory-scale, log-magnitude spectrograms. Beyond that, almost anything will probably work quite well.
